Refs QA-3629 Refine behavior of matching excel fields to deprecated p…

kmtracey opened the pull request:
…roperties

Do not list deprecated properties in dropdown, but allow them to be typed and auto-suggested
When a deprecated property has been typed/suggested, show a warning that it is a deprecated property
Also show any formats/valid values in this case
Uncheck the "Create new property" in this case as well, since the property exists

In addition, this PR sanitizes allowed values input into the data dictionary before sending to the backend.
(Having script tags (for example) in the backend for suggested valid values causes the importer matching page
to not work as expected -- it seems the js initial data doesn't get deserialied properly -- so stripping in
the same way as is done for display on data dictiony seemed easist to fix this oddity.)

Finally, as requested by Daniel, disable the Next button on the initial upload page when there is no file
chosen.

[safety info removed because this isn't a PR into master]

kmtracey committed:
Refs QA-3629 Refine behavior of matching excel fields to deprecated properties

Do not list deprecated properties in dropdown, but allow them to be typed and auto-suggested
When a deprecated property has been typed/suggested, show a warning that it is a deprecated property
Also show any formats/valid values in this case
Uncheck the "Create new property" in this case as well, since the property exists

In addition, this PR sanitizes allowed values input into the data dictionary before sending to the backend.
(Having script tags (for example) in the backend for suggested valid values causes the importer matching page
to not work as expected -- it seems the js initial data doesn't get deserialied properly -- so stripping in
the same way as is done for display on data dictiony seemed easist to fix this oddity.)

Finally, as requested by Daniel, disable the Next button on the initial upload page when there is no file
chosen.

kmtracey committed:
Refs QA-3629 js fixes per sticklerbot

orangejenny commented on corehq/apps/case_importer/static/case_importer/js/main.js:
```diff
@@ -26,6 +26,19 @@ hqDefine("case_importer/js/main", [
+            if (fileName) {
+                $(':submit').attr('disabled', false).removeClass('disabled');
+            } else {
+                $(':submit').attr('disabled', true).addClass('disabled');
```
You can also use `$(':submit').disableButtonNoSpinner()` for readability (and `enableButton` above), but it's the same code.

kmtracey replied:
Ah, didn't find those in a brief look around. Updated to use them instead.

orangejenny commented on corehq/apps/data_dictionary/util.py:
```diff
@@ -196,7 +204,7 @@ def get_data_dict_props_by_case_type(domain):
         case_type: {prop.name for prop in props} for case_type, props in groupby(
             CaseProperty.objects
-            .filter(case_type__domain=domain, deprecated=False)
+            .filter(case_type__domain=domain)
```
What's the purpose of this change? This function is called in the [ParentCasePropertyBuilder](https://github.com/dimagi/commcare-hq/blob/06f4aa6744babf691ed5b0a42049cb6856378d3b/corehq/apps/app_manager/app_schemas/case_properties.py#L373), which makes me think it might make deprecated properties show up in places beyond the case importer, such as the "App Properties" window that shows up when you're building a form.

kmtracey replied:
Oops I think I forgot to track down if this was used elsewhere. I've updated the PR to include a new param that allows for return of deprecated properties when desired (the f/e here wants to warn when one is typed or suggested, and also to show the valid values and formats even for deprecated ones).

orangejenny commented on corehq/apps/case_importer/suggested_fields.py:
```diff
@@ -33,11 +33,14 @@ def _combine_field_specs(field_specs, exclude_fields):
 def get_suggested_case_fields(domain, case_type, exclude=None):
     exclude_fields = set(RESERVED_FIELDS) | set(exclude or [])
     hints_dict = get_values_hints_dict(domain, case_type)
+    deprecated_fields = get_deprecated_fields(domain, case_type)
```
I think this should check the DATA_DICTIONARY flag is on (either here or in the function).

kmtracey replied:
Hmm, should the `get_values_hints_dict` also be suppressed if data dictonary is not enabled? For that matter should the whole column for "Valid values and formats" be suppressed when this flag is not on?

orangejenny replied:
Good point. Yeah, the page shouldn't show the column if the feature isn't turned on. Turning it off at the UI level then probably makes it not really matter if `get_values_hints_dict` and `deprecated_fields` are populated here.

orangejenny reviewed (commented):
Just some minor feedback/questions.

kmtracey committed:
Refs QA-3629 use friendly-named utils for button enable/disable,

also limit return of deprecated properties by utility function to code requesting deprecated properties.

orangejenny commented:
The point about hiding the column if the feature flag is disabled is good, but since it doesn't affect this fix directly, I'm going to merge this so QA can verify.

kmtracey commented:
I can hopefully put up a pr to suppress the column first thing tomorrow...
